File: EntreNosApp/src/ProjetoEncontros.Dominio/Notificacoes/NotificacaoDoUsuario.cs
using ProjetoEncontros.Dominio.Compartilhado;

namespace ProjetoEncontros.Dominio.Notificacoes;

public sealed class NotificacaoDoUsuario : Entidade
{
    public const int TamanhoMaximoDoTitulo = 120;
    public const int TamanhoMaximoDaMensagem = 300;
    public const int TamanhoMaximoDaChaveDeIdempotencia = 100;

    private NotificacaoDoUsuario()
    {
        Titulo = "Notificação";
        Mensagem = "Você tem uma nova notificação.";
    }

    private NotificacaoDoUsuario(
        Guid identificador,
        Guid identificadorDoUsuario,
        TipoDeNotificacao tipo,
        string titulo,
        string mensagem,
        Guid? identificadorDoEncontro,
        Guid? identificadorDoConvite,
        Guid? identificadorDoItem,
        DateTimeOffset criadaEm,
        string? chaveDeIdempotencia)
        : base(identificador, criadaEm)
    {
        if (identificadorDoUsuario == Guid.Empty)
        {
            throw new ExcecaoDeDominioException("O identificador do usuário da notificação não pode ser vazio.");
        }

        ValideIdentificadorOpcional(identificadorDoEncontro, "O identificador do encontro da notificação não pode ser vazio.");
        ValideIdentificadorOpcional(identificadorDoConvite, "O identificador do convite da notificação não pode ser vazio.");
        ValideIdentificadorOpcional(identificadorDoItem, "O identificador do item da notificação não pode ser vazio.");

        IdentificadorDoUsuario = identificadorDoUsuario;
        Tipo = tipo;
        Titulo = NormalizeTexto(titulo, TamanhoMaximoDoTitulo, "O título da notificação é obrigatório.", "O título da notificação não pode ultrapassar 120 caracteres.");
        Mensagem = NormalizeTexto(mensagem, TamanhoMaximoDaMensagem, "A mensagem da notificação é obrigatória.", "A mensagem da notificação não pode ultrapassar 300 caracteres.");
        IdentificadorDoEncontro = identificadorDoEncontro;
        IdentificadorDoConvite = identificadorDoConvite;
        IdentificadorDoItem = identificadorDoItem;
        ChaveDeIdempotencia = NormalizeTextoOpcional(
            chaveDeIdempotencia,
            TamanhoMaximoDaChaveDeIdempotencia,
            "A chave de idempotência da notificação não pode ultrapassar 100 caracteres.");
        Situacao = SituacaoDaNotificacao.NaoLida;
    }

    public Guid IdentificadorDoUsuario { get; private set; }

    public TipoDeNotificacao Tipo { get; private set; }

    public string Titulo { get; private set; }

    public string Mensagem { get; private set; }

    public Guid? IdentificadorDoEncontro { get; private set; }

    public Guid? IdentificadorDoConvite { get; private set; }

    public Guid? IdentificadorDoItem { get; private set; }

    public string? ChaveDeIdempotencia { get; private set; }

    public SituacaoDaNotificacao Situacao { get; private set; }

    public DateTimeOffset? LidaEm { get; private set; }

    public bool EstaLida
    {
        get
        {
            return Situacao == SituacaoDaNotificacao.Lida;
        }
    }

    public bool EstaNaoLida
    {
        get
        {
            return Situacao == SituacaoDaNotificacao.NaoLida;
        }
    }

    public static NotificacaoDoUsuario Crie(
        Guid identificador,
        Guid identificadorDoUsuario,
        TipoDeNotificacao tipo,
        string titulo,
        string mensagem,
        Guid? identificadorDoEncontro,
        Guid? identificadorDoConvite,
        Guid? identificadorDoItem,
        DateTimeOffset criadaEm,
        string? chaveDeIdempotencia = null)
    {
        return new(
            identificador,
            identificadorDoUsuario,
            tipo,
            titulo,
            mensagem,
            identificadorDoEncontro,
            identificadorDoConvite,
            identificadorDoItem,
            criadaEm,
            chaveDeIdempotencia);
    }

    public void MarqueComoLida(DateTimeOffset lidaEm)
    {
        if (EstaLida)
        {
            return;
        }

        Situacao = SituacaoDaNotificacao.Lida;
        LidaEm = lidaEm;
    }

    private static string NormalizeTexto(string texto, int tamanhoMaximo, string mensagemObrigatoria, string mensagemDeTamanho)
    {
        if (string.IsNullOrWhiteSpace(texto))
        {
            throw new ExcecaoDeDominioException(mensagemObrigatoria);
        }

        string textoNormalizado = texto.Trim();

        if (textoNormalizado.Length > tamanhoMaximo)
        {
            throw new ExcecaoDeDominioException(mensagemDeTamanho);
        }

        return textoNormalizado;
    }

    private static void ValideIdentificadorOpcional(Guid? identificador, string mensagem)
    {
        if (identificador.HasValue && identificador.Value == Guid.Empty)
        {
            throw new ExcecaoDeDominioException(mensagem);
        }
    }

    private static string? NormalizeTextoOpcional(
        string? texto,
        int tamanhoMaximo,
        string mensagemDeTamanho)
    {
        if (string.IsNullOrWhiteSpace(texto))
        {
            return null;
        }

        string textoNormalizado = texto.Trim();

        if (textoNormalizado.Length > tamanhoMaximo)
        {
            throw new ExcecaoDeDominioException(mensagemDeTamanho);
        }

        return textoNormalizado;
    }
}

File: EntreNosApp/src/ProjetoEncontros.AplicativoWeb/lib/funcionalidades/notificacoes/modelos/notificacao_do_usuario.dart
class NotificacaoDoUsuario {
  const NotificacaoDoUsuario({
    required this.identificador,
    required this.tipo,
    required this.titulo,
    required this.mensagem,
    required this.situacao,
    required this.criadaEm,
    this.identificadorDoEncontro,
    this.identificadorDoConvite,
    this.identificadorDoItem,
    this.lidaEm,
  });

  factory NotificacaoDoUsuario.deJson(Map<String, dynamic> json) {
    String? lidaEm = json['lidaEm'] as String?;

    return NotificacaoDoUsuario(
      identificador: json['identificador'] as String,
      tipo: json['tipo'] as String,
      titulo: json['titulo'] as String,
      mensagem: json['mensagem'] as String,
      identificadorDoEncontro: json['identificadorDoEncontro'] as String?,
      identificadorDoConvite: json['identificadorDoConvite'] as String?,
      identificadorDoItem: json['identificadorDoItem'] as String?,
      situacao: json['situacao'] as String,
      criadaEm: DateTime.parse(json['criadaEm'] as String).toLocal(),
      lidaEm: lidaEm == null ? null : DateTime.parse(lidaEm).toLocal(),
    );
  }

  final String identificador;
  final String tipo;
  final String titulo;
  final String mensagem;
  final String? identificadorDoEncontro;
  final String? identificadorDoConvite;
  final String? identificadorDoItem;
  final String situacao;
  final DateTime criadaEm;
  final DateTime? lidaEm;

  bool get estaLida => lidaEm != null || situacao.toLowerCase() == 'lida';
}

class ListaDeNotificacoes {
  const ListaDeNotificacoes({
    required this.quantidadeNaoLida,
    required this.notificacoes,
  });

  factory ListaDeNotificacoes.deJson(Map<String, dynamic> json) {
    List<dynamic> itens = json['notificacoes'] as List<dynamic>? ?? <dynamic>[];

    return ListaDeNotificacoes(
      quantidadeNaoLida: json['quantidadeNaoLida'] as int? ?? 0,
      notificacoes: itens
          .map(
            (dynamic item) => NotificacaoDoUsuario.deJson(
              Map<String, dynamic>.from(item as Map<dynamic, dynamic>),
            ),
          )
          .toList(),
    );
  }

  final int quantidadeNaoLida;
  final List<NotificacaoDoUsuario> notificacoes;
}
